docs/proposals/kubelet-tls-bootstrap.md

```diff
@@ -0,0 +1,231 @@
+
+We introduce a new API object to represent PKCS#10 certificate signing
+requests. It will be accessible under:
+
+`/api/vX/certificaterequests/mycsr`
+
+It will have the following structure:
+
+```go
+// Describes a certificate signing request
+type CertificateSigningRequest struct {
+        api.TypeMeta         `json:",inline"`
+        api.ObjectMeta       `json:"metadata,omitempty"`
+
+        // Specifies the behavior of the CSR
+        Spec CertificateSigningRequestSpec
+
+        // Most recently observed status of the CSR
+        Status CertificateSigningRequestStatus
+}
+
+type CertificateSigningRequestSpec struct {
+        // Raw PKCS#10 CSR data
+        CertificateRequest []byte
+
+        // Fingerprint of the public key that signed the CSR
+        Fingerprint string
+
+        // Subject fields from the CSR
+        Subject pkix.Name
+
+        // DNS SANs from the CSR
+        Hostnames []string
+
+        // IP SANs from the CSR
+        IPAddresses []string
+
+        // Extra information the node wishes to send with the request
+        ExtraInfo []string
+}
+
+type CertificateSigningRequestStatus struct {
+        // Indicates whether CSR has a response yet. Default is Unknown. Status
+        // is True for approval and False for rejections.
+        Status api.ConditionStatus
+
+        // If CSR was rejected, these contain the reason why (if any was supplied).
+        Reason string
+        Message string
+
+        // If CSR was approved, this contains the issued certificate.
+        Certificate []byte
+}
+```
+
+We also introduce CertificateSigningRequestList to allow listing all the CSRs in the cluster:
+
+```go
+type CertificateSigningRequestList struct {
+        api.TypeMeta
+        api.ListMeta
+
+        Items []CertificateSigningRequest
+}
+```
+
+## Certificate Request Process
+
+### Node intialization
+
+When the kubelet executes it checks a location on disk for TLS assets
+(currently `/var/run/kubernetes/kubelet.{key,crt}` by default). If it finds
+them, it proceeds. If there are no TLS assets, the kubelet generates a keypair
+and self-signed certificate. We propose the following optional fallback behavior:
+
+1. Generate a keypair
+2. Generate a CSR for that keypair with CN set to the hostname (or
+   `--hostname-override` value) and DNS/IP SANs supplied with whatever values
+   the host knows for itself.
+3. Post the CSR to the CSR API endpoint.
+4. Set a watch on the CSR object to be notified of approval or rejection.
+
+### Controller response
+
+The apiserver must first validate the signature on the raw CSR data and reject
+requests featuring invalid CSRs. It then persists the
+CertificateSigningRequests and exposes the List of all CSRs for an
+administrator to approve or reject. The apiserver should watch for updates the
+Status field of any CertificateSigningRequest. When a CSR is approved
+(signified by Status changing from Unknown to True) the apiserver should
```

Comment:
After trying both options, I agree that the controller approach is better here. So it will be a straight update from the controller, but I think we should still stick with a simpler status. I think the right three are:

1. Submitted
2. Approved and available
3. Denied

What would "approved-but-not-yet-signed" mean to a consumer of this API?